File: code/mergesort.py
# -*- coding: utf-8 -*-
import math


def merge(A, p, q, r):
    """
    Sorts A by the merge sort algorithm.
    Uses 1-based indexing.
    This Python implementation is designed to be as close to the pseudocode as possible. It is not elegant Python code.
    :param A: List of numbers to be sorted.
    :param p: The index of the first element in the part to be sorted.
    :param q: Middlepoint in the part to be sorted.
    :param r: The index of the last element in the part to be sorted.
    """
    n1 = q - p + 1
    n2 = r - q
    
    L = [0] * n1
    R = [0] * n2
    
    for i in list(range(n1)):
        L[i] = A[p + i - 1]
    
    for j in list(range(n2)):
        R[j] = A[q + j]
    
    L.append(float('inf'))
    R.append(float('inf'))
    
    i = 1 - 1 # Subtract 1 to adjust to Python indexing
    j = 1 - 1 # Subtract 1 to adjust to Python indexing
    
    for k in list(range(p - 1, r)): # Subtract 1 from q to adjust to Python range object
        if L[i] <= R[j]:
            A[k] = L[i]
            i = i + 1
        else:
            A[k] = R[j]
            j = j + 1


def mergesort(A, p, r):
    """
    Halves the list recursively and calls the merge() function
    :param A: List of numbers to be sorted.
    :param p: The index of the first element in the part to be sorted.
    :param r: The index of the last element in the part to be sorted.
    """
    if p < r:
        q = math.floor((p+r)/2)
        merge_sort(A, p, q)
        merge_sort(A, q+1, r)
        merge(A, p, q, r)

File: code/c/mergesort.c
#include<stdio.h>
#include<stdlib.h>
#include<math.h>

void merge(int* array, long p, long r, long q){
	long n1 = r-p+1;
	long n2 = q - r;

	int* L = malloc(sizeof(int)*(n1 + 1));
	int* R = malloc(sizeof(int)*(n2 + 1));
	
	for(long i = 0; i < n1; i ++) {
		*(L + i) = *(array + p + i -1);
	}
	for(long i = 0; i < n2; i ++) {
		*(R + i) = *(array + q + i);
	}



	free(L);
	free(R);
}

void mergesort(int* array, long p, long q){
	if(p < q){
		long r = floor((p+q)/2);
		mergesort(array,p,r);
		mergesort(array,r+1,q);
		merge(array,p,q,r);
	}
}

int main(){
	long size = pow(2,10);
	int* ptr = malloc(size);
	for(long i = 0; i < size; i++){
		*(ptr + i) = (int)(rand() % 10000);
	}
	mergesort(ptr,0,size-1);
}
